test/gtest/ucp/test_ucp_reconfigure.cc

```diff
@@ -0,0 +1,264 @@
+/**
+* Copyright (c) NVIDIA CORPORATION & AFFILIATES, 2001-2024. ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#include "ucp_test.h"
+#include "common/test.h"
+
+extern "C" {
+#include <ucp/core/ucp_ep.inl>
+#include <ucp/wireup/address.h>
+#include <ucp/wireup/wireup_ep.h>
+}
+
+class test_ucp_reconfigure : public ucp_test {
+protected:
+    using address_t = std::pair<void*, ucp_unpacked_address_t>;
+    using address_p = std::unique_ptr<address_t, void (*)(address_t*)>;
+
+    class entity {
+    public:
+        void connect(const entity &other, bool is_exclude_iface);
+
+        void verify_configuration(const entity &other) const;
+
+        bool is_reconfigured() const
+        {
+            return m_cfg_index != ep()->cfg_index;
+        }
+
+        void set_worker_entity(const ucp_test_base::entity *e)
+        {
+            m_worker_entity = e;
+        }
+
+        ucp_ep_h ep() const
+        {
+            return m_ep;
+        }
+
+        ucp_worker_h worker() const
+        {
+            return m_worker_entity->worker();
+        }
+
+        void cleanup()
+        {
+            if (m_ep != nullptr) {
+                ucp_ep_destroy(m_ep);
+            }
+        }
+
+    private:
+        void store_config();
+        ucp_tl_bitmap_t ep_tl_bitmap() const;
+        address_p get_address(bool ep_only) const;
+        bool has_matching_lane(ucp_ep_h ep, ucp_lane_index_t lane_idx,
+                               const entity &other) const;
+
+        ucp_worker_cfg_index_t       m_cfg_index     = UCP_WORKER_CFG_INDEX_NULL;
+        std::vector<uct_ep_h>        m_uct_eps;
+        const ucp_test_base::entity *m_worker_entity = nullptr;
+        ucp_ep_h                     m_ep            = nullptr;
+    };
+
+    void create_entities_and_connect()
+    {
+        m_sender.set_worker_entity(create_entity(true));
+        m_receiver.set_worker_entity(create_entity(false));
+
+        m_sender.connect(m_receiver, is_exclude_iface());
+        m_receiver.connect(m_sender, is_exclude_iface());
+    }
+
+    void cleanup()
+    {
+        m_sender.cleanup();
+        m_receiver.cleanup();
+        ucp_test::cleanup();
+    }
+
+    entity m_sender;
+    entity m_receiver;

```

Comment:
can `entities_storage` class be reused?

Reply:
IIUC, it cannot be used because `test_ucp_reconfigure::entity` does not derive from `ucp_test_base::entity`

Reply:
1) BTW, what's the reason to add new entity class? do we really need it?
2) `entities_storage` is a template, so it can be used with any type of entities.

Reply:
1. it's required for controlling tl_bitmap which creates asymmetric configs in local vs remote.
2. `entities_storage` is already used by `ucp_test` which is the base class of `test_ucp_reconfigure`, so `test_ucp_reconfigure` cannot add another `entities_storage` with a different entity type

Reply:
1. can existing entity class be extended with missed functionality?
2. so we have common `entities_storage` and local members of completely different type of entities, IMO need to reuse base class since there is a set of methods which is based on common entities storage and widely used in tests, like is_loopback(), skip_<smth>(), etc

Reply:
1. IMO it's better not to extend the existing entity class only to support this specific case. better it's all contained in test_reconfigure.
2. OK